Implement new InterruptNumber trait from next cortex-m release

therealprof opened the pull request:
This eliminates the bare-metal and cortex-m-rt dependencies for Cortex-M
and thus changes CI and svd2rust-regress accordingly.

Replaces #455

Signed-off-by: Daniel Egger <daniel@eggers-club.de>

rust-highfive commented:
r? @adamgreig

(rust_highfive has picked a reviewer for you, use r? to override)

adamgreig commented:
This LGTM to me, except I don't understand why we were using `cortex_m_rt::interrupt` before anyway.

therealprof commented:
I have no idea why we're not getting random errors in for different vendors. Need to investigate after cooking...

therealprof commented:
@adamgreig If we're happy with this we'll only need the cortex-m release to finalize this PR.

adamgreig commented on src/generate/device.rs:
```diff
@@ -112,7 +112,7 @@ pub fn render(
         out.extend(quote! {
             pub use cortex_m::peripheral::Peripherals as CorePeripherals;
             #[cfg(feature = "rt")]
-            pub use cortex_m_rt::interrupt;
```
I think this was here to ensure the svd2rust generated PAC exports an `interrupt` macro, since the cortex-m-rt documentation says you should use `interrupt` from your PAC and not from c-m-rt. So, I think we should keep the `pub use cortex_m_rt::interrupt`, but we don't need to `pub use cortex_m::interrupt::InterruptNumber` at all, since the impl writes the full path anyway.

adamgreig reviewed (approved):
Looks good. I think we'll have to try it out a bit with cortex-m and PACs to be sure it's working as we expect but PR looks fine to me. Could you add a CHANGELOG entry?

therealprof commented:
@adamgreig I've updated the PR to use the release and added a CHANGELOG. I also ran `svd2rust-regress` locally but I haven't tried adapting a HAL impl yet.

adamgreig reviewed (approved):
bors merge

Rahix commented on src/generate/interrupt.rs:
```diff
@@ -180,9 +180,9 @@ pub fn render(
             }
 
-            unsafe impl bare_metal::Nr for Interrupt {
+            unsafe impl cortex_m::interrupt::InterruptNumber for Interrupt {
```
This change breaks the build for [`avr-device`](https://github.com/Rahix/avr-device).  The problem is that you modified the `else` branch which is for everything but `msp430` ...

I suppose this code needs to be refactored such that this line is only emitted for `cortex-m`.  I would assume other platforms (like `riscv`) will also run into the same issue.

Cc: @trembel

therealprof commented:
@Rahix Thanks for noticing. Indeed we has some discussion in #455 about other architectures and the approach to be taken there.

Since this doesn't break RISC-V in CI, this would imply that either it doesn't break or that it is not exercised properly.

AVR is completely disabled in CI which is why there's no indication of breakage: https://github.com/rust-embedded/svd2rust/blob/c68b71c09c87ff17270e9e8e6cb63f45276b5068/ci/script.sh#L78

Rahix commented:
> Since this doesn't break RISC-V in CI, this would imply that either it doesn't break or that it is not exercised properly.

Hmm, from reading the code in `interrupt.rs` I cannot tell how RISC-V manages to evade this problem (unless all tested SVD files contain no interrupts at all which I would assume is not the case).  Would need another closer look ...

> AVR is not even tested in CI which would be a good first thing to add.

I can look into adding that.  For AVR, the vendor files are in a format called ATDF which we first convert to SVD using a small utility called [`atdf2vd`](https://github.com/Rahix/atdf2svd).  I suppose I should upload a pre-generated SVD somewhere and reference that for the CI tests?
